File: python/csv-to-json.py
#!/usr/bin/env python3
"""
CSV to JSON Config Converter

This script converts a CSV file containing skill information to a JSON configuration file.
It automatically detects and converts file encoding to UTF-8.
"""

import os
import sys
import csv
import json
import argparse
from datetime import datetime


def write_log(message):
    """Write log message with timestamp"""
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{timestamp}] {message}")


def get_file_encoding(file_path):
    """Detect file encoding"""
    try:
        with open(file_path, 'rb') as f:
            bytes_data = f.read(4)
        
        if len(bytes_data) >= 3 and bytes_data[0] == 0xEF and bytes_data[1] == 0xBB and bytes_data[2] == 0xBF:
            return "UTF8"
        if len(bytes_data) >= 2 and bytes_data[0] == 0xFF and bytes_data[1] == 0xFE:
            return "Unicode"
        if len(bytes_data) >= 2 and bytes_data[0] == 0xFE and bytes_data[1] == 0xFF:
            return "BigEndianUnicode"
        return "Default"
    except Exception:
        return "Locked"


def convert_to_utf8(file_path):
    """Convert file to UTF-8 encoding"""
    encoding = get_file_encoding(file_path)
    write_log(f"Detected encoding: {encoding}")
    
    if encoding == "UTF8":
        write_log("File is already UTF-8, no conversion needed")
        return True
    
    if encoding == "Locked":
        write_log("WARNING: File is locked by another process, skipping encoding conversion")
        write_log("Please close the file in other applications and try again if encoding issues occur")
        return True
    
    write_log(f"Converting file from {encoding} to UTF-8...")
    
    try:
        if encoding == "Default":
            # 尝试使用常见编码进行解码
            encodings_to_try = ['utf-8', 'gbk', 'gb2312', 'latin-1']
            content = None
            used_encoding = None
            
            for enc in encodings_to_try:
                try:
                    with open(file_path, 'r', encoding=enc) as f:
                        content = f.read()
                    # 简单验证内容是否有效
                    if content and (any(c.isalnum() for c in content) or len(content) > 0):
                        used_encoding = enc
                        write_log(f"Successfully decoded with encoding: {enc}")
                        break
                except UnicodeDecodeError:
                    # 忽略解码错误，尝试下一个编码
                    pass
            
            if not content:
                # 如果所有编码都失败，使用 'replace' 模式
                with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
                    content = f.read()
                used_encoding = 'utf-8 (with replace)'
                write_log("Using utf-8 with replace as fallback")
        else:
            # 使用检测到的编码
            with open(file_path, 'r', encoding=encoding.lower() if encoding != "Default" else 'utf-8', errors='replace') as f:
                content = f.read()
        
        # Write with UTF-8
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(content)
        
        write_log("Conversion completed successfully")
        return True
    except Exception as e:
        write_log(f"ERROR: Failed to convert encoding: {e}")
        return False


def main():
    """Main function"""
    parser = argparse.ArgumentParser(description='Convert CSV to JSON config file')
    parser.add_argument('-CsvFile', default='skills.csv', help='CSV file path')
    parser.add_argument('-JsonFile', default='sync-config.json', help='JSON file path')
    parser.add_argument('-TargetPath', default='C:\\Users\\admin\\.trae-cn\\skills\\', help='Target path for skills')
    parser.add_argument('-OutputType', default='sync', choices=['sync', 'init'], help='Output type: sync or init')
    args = parser.parse_args()
    
    csv_file = args.CsvFile
    json_file = args.JsonFile
    target_path = args.TargetPath
    output_type = args.OutputType
    
    # Resolve full path
    csv_full_path = os.path.abspath(csv_file)
    
    if not os.path.exists(csv_full_path):
        write_log(f"ERROR: CSV file '{csv_full_path}' does not exist")
        sys.exit(1)
    
    # Convert to UTF-8
    if not convert_to_utf8(csv_full_path):
        write_log("ERROR: Failed to ensure UTF-8 encoding for CSV file")
        sys.exit(1)
    
    write_log(f"Reading CSV file: {csv_full_path}")
    
    # Read CSV
    csv_data = []
    with open(csv_full_path, 'r', encoding='utf-8-sig') as f:  # 使用 utf-8-sig 处理 BOM
        reader = csv.DictReader(f)
        for row in reader:
            csv_data.append(row)
    
    if not csv_data:
        write_log("WARNING: CSV file is empty")
        sys.exit(0)
    
    write_log(f"CSV file contains {len(csv_data)} rows")
    
    if output_type == 'init':
        # 处理 init 模式：生成 init-config.json
        write_log("Processing in init mode: generating init-config.json")
        
        # 从 CSV 数据中提取唯一的仓库信息
        repo_map = {}
        for row in csv_data:
            repo_path = row.get('repoPath')
            repo_name = row.get('repoName')
            local_path = row.get('localPath')
            repo_url = row.get('repoUrl')
            
            if not repo_path:
                write_log("Skipping empty repoPath")
                continue
            
            if repo_path not in repo_map:
                repo_map[repo_path] = {
                    'repoName': repo_name,
                    'localPath': local_path,
                    'repoUrl': repo_url
                }
        
        # 生成 init-config.json 格式
        init_config = {
            'repos': []
        }
        
        for key, value in repo_map.items():
            repo_url = value.get('repoUrl')
            
            # 如果 CSV 中没有提供 repoUrl，则从 repoPath 生成
            if not repo_url:
                # 从 repoPath 提取 GitHub 仓库信息
                parts = key.split('\\')
                if len(parts) >= 2:
                    owner = parts[-3] if len(parts) >= 3 else ''
                    repo = parts[-2]
                    repo_url = f"https://github.com/{owner}/{repo}.git"
                else:
                    repo_url = ""
            
            repo_config = {
                'repoName': value['repoName'],
                'repoUrl': repo_url,
                'localPath': value['localPath']
            }
            init_config['repos'].append(repo_config)
        
        # 保存到文件
        with open(json_file, 'w', encoding='utf-8') as f:
            json.dump(init_config, f, indent=2, ensure_ascii=False)
        write_log(f"Init config saved to: {json_file}")
        write_log("Config content:")
        print(json.dumps(init_config, indent=2, ensure_ascii=False))
    else:
        # 处理 sync 模式：生成 sync-config.json
        write_log("Processing in sync mode: generating sync-config.json")
        
        # Read existing JSON
        json_config = []
        if os.path.exists(json_file):
            write_log(f"Reading existing JSON file: {json_file}")
            with open(json_file, 'r', encoding='utf-8') as f:
                try:
                    parsed_json = json.load(f)
                    if isinstance(parsed_json, list):
                        json_config = parsed_json
                    else:
                        json_config = [parsed_json]
                except json.JSONDecodeError:
                    write_log("WARNING: Invalid JSON file, creating new configuration")
                    json_config = []
        else:
            write_log("JSON file does not exist, creating new configuration")
        
        # Find or create target group
        target_group = None
        for group in json_config:
            if group.get('targetPath') == target_path:
                target_group = group
                break
        
        if not target_group:
            write_log(f"Creating new target group: {target_path}")
            target_group = {
                'targetPath': target_path,
                'repos': []
            }
            json_config = [target_group]
        
        # Create repo map
        repo_map = {}
        repo_name_map = {}
        for repo in target_group.get('repos', []):
            repo_map[repo['repoPath']] = repo['skillNames']
            if 'repoName' in repo:
                repo_name_map[repo['repoPath']] = repo['repoName']
        
        modified = False
        
        # Process CSV data
        for row in csv_data:
            repo_path = row.get('repoPath')
            skill_name = row.get('skillName')
            repo_name = row.get('repoName')
            
            if not repo_path or not skill_name:
                write_log(f"Skipping empty row: repoPath={repo_path}, skillName={skill_name}")
                continue
            
            if repo_path in repo_map:
                if skill_name not in repo_map[repo_path]:
                    write_log(f"Adding skill '{skill_name}' to existing repo: {repo_path}")
                    repo_map[repo_path].append(skill_name)
                    modified = True
                else:
                    write_log(f"Skill '{skill_name}' already exists in repo: {repo_path}")
                
                # Keep repoName unchanged from existing config
                # Do not update repoName from CSV
            else:
                write_log(f"Creating new repo config: {repo_path}, adding skill: {skill_name}")
                repo_map[repo_path] = [skill_name]
                if repo_name:
                    repo_name_map[repo_path] = repo_name
                modified = True
        
        # Update repos
        new_repos = []
        for key, value in repo_map.items():
            repo_config = {
                'repoPath': key,
                'skillNames': value
            }
            
            if key in repo_name_map:
                repo_config['repoName'] = repo_name_map[key]
            
            new_repos.append(repo_config)
        
        target_group['repos'] = new_repos
        
        # Save JSON
        if modified:
            output_config = json_config[0] if len(json_config) == 1 else json_config
            with open(json_file, 'w', encoding='utf-8') as f:
                json.dump(output_config, f, indent=2, ensure_ascii=False)
            write_log(f"JSON config saved to: {json_file}")
            write_log("Config content:")
            print(json.dumps(output_config, indent=2, ensure_ascii=False))
        else:
            write_log("No updates needed")
    
    write_log("Processing completed")


if __name__ == '__main__':
    main()

File: go/csv-to-json.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// 全局变量
var (
	dateFormat = "2006-01-02 15:04:05"
)

// 初始化配置结构
type InitConfig struct {
	Repos []RepoConfig `json:"repos"`
}

// 仓库配置结构
type RepoConfig struct {
	RepoName  string `json:"repoName,omitempty"`
	RepoUrl   string `json:"repoUrl"`
	LocalPath string `json:"localPath,omitempty"`
}

// 同步配置结构
type SyncConfig struct {
	TargetPath string       `json:"targetPath"`
	Repos      []SyncRepoConfig `json:"repos"`
}

// 同步仓库配置结构
type SyncRepoConfig struct {
	RepoPath   string   `json:"repoPath"`
	RepoName   string   `json:"repoName,omitempty"`
	SkillNames []string `json:"skillNames"`
}

func main() {
	// 解析命令行参数
	csvFile := "skills.csv"
	jsonFile := "sync-config.json"
	targetPath := "C:\\Users\\admin\\.trae-cn\\skills\\"
	outputType := "sync" // 默认值为 sync

	for i := 0; i < len(os.Args); i++ {
		if os.Args[i] == "-CsvFile" && i+1 < len(os.Args) {
			csvFile = os.Args[i+1]
		} else if os.Args[i] == "-JsonFile" && i+1 < len(os.Args) {
			jsonFile = os.Args[i+1]
		} else if os.Args[i] == "-TargetPath" && i+1 < len(os.Args) {
			targetPath = os.Args[i+1]
		} else if os.Args[i] == "-OutputType" && i+1 < len(os.Args) {
			outputType = os.Args[i+1]
		}
	}

	writeLog("Starting CSV to JSON conversion...")

	// 解析完整路径
	csvPath, err := filepath.Abs(csvFile)
	if err != nil {
		writeLog(fmt.Sprintf("ERROR: Failed to get absolute path: %v", err))
		os.Exit(1)
	}

	// 检查CSV文件是否存在
	if _, err := os.Stat(csvPath); os.IsNotExist(err) {
		writeLog(fmt.Sprintf("ERROR: CSV file '%s' does not exist", csvPath))
		os.Exit(1)
	}

	writeLog(fmt.Sprintf("Reading CSV file: %s", csvPath))

	// 读取CSV文件
	csvData, err := readCsvFile(csvPath)
	if err != nil {
		writeLog(fmt.Sprintf("ERROR: Failed to read CSV file: %v", err))
		os.Exit(1)
	}

	if len(csvData) == 0 {
		writeLog("WARNING: CSV file is empty")
		os.Exit(0)
	}

	writeLog(fmt.Sprintf("CSV file contains %d rows", len(csvData)))

	jsonPath, err := filepath.Abs(jsonFile)
	if err != nil {
		writeLog(fmt.Sprintf("ERROR: Failed to get absolute path: %v", err))
		os.Exit(1)
	}

	if outputType == "init" {
		// 处理init模式：生成init-config.json
		writeLog("Processing in init mode: generating init-config.json")
		processInitMode(csvData, jsonPath)
	} else {
		// 处理sync模式：生成sync-config.json
		writeLog("Processing in sync mode: generating sync-config.json")
		processSyncMode(csvData, jsonPath, targetPath)
	}

	writeLog("Processing completed")
}

// 读取CSV文件
func readCsvFile(csvPath string) ([]map[string]string, error) {
	file, err := os.Open(csvPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	reader.TrimLeadingSpace = true

	// 读取表头
	headers, err := reader.Read()
	if err != nil {
		return nil, err
	}

	// 处理UTF-8 BOM
	if len(headers) > 0 && strings.HasPrefix(headers[0], "\ufeff") {
		headers[0] = strings.TrimPrefix(headers[0], "\ufeff")
	}

	// 读取数据
	csvData := []map[string]string{}
	for {
		row, err := reader.Read()
		if err != nil {
			break
		}

		rowData := make(map[string]string)
		for i := 0; i < len(headers) && i < len(row); i++ {
			rowData[headers[i]] = strings.TrimSpace(row[i])
		}
		csvData = append(csvData, rowData)
	}

	return csvData, nil
}

// 处理init模式
func processInitMode(csvData []map[string]string, jsonPath string) {
	// 从CSV数据中提取唯一的仓库信息
	repoMap := make(map[string]map[string]string)
	for _, row := range csvData {
		repoPath := strings.TrimSpace(row["repoPath"])
		repoName := strings.TrimSpace(row["repoName"])
		localPath := strings.TrimSpace(row["localPath"])
		repoUrl := strings.TrimSpace(row["repoUrl"])

		if repoPath == "" {
			writeLog("Skipping empty repoPath")
			continue
		}

		if _, ok := repoMap[repoPath]; !ok {
			repoInfo := make(map[string]string)
			if repoName != "" {
				repoInfo["repoName"] = repoName
			}
			if localPath != "" {
				repoInfo["localPath"] = localPath
			}
			if repoUrl != "" {
				repoInfo["repoUrl"] = repoUrl
			}
			repoMap[repoPath] = repoInfo
		}
	}

	// 生成init-config.json格式
	initConfig := InitConfig{}
	initConfig.Repos = []RepoConfig{}

	for repoPath, repoInfo := range repoMap {
		repoUrl := repoInfo["repoUrl"]

		// 如果CSV中没有提供repoUrl，则从repoPath生成
		if repoUrl == "" {
			// 从repoPath提取GitHub仓库信息
			parts := strings.Split(repoPath, "\\")
			if len(parts) >= 2 {
				owner := ""
				if len(parts) >= 3 {
					owner = parts[len(parts)-3]
				}
				repo := parts[len(parts)-2]
				repoUrl = fmt.Sprintf("https://github.com/%s/%s.git", owner, repo)
			}
		}

		repoConfig := RepoConfig{
			RepoName:  repoInfo["repoName"],
			RepoUrl:   repoUrl,
			LocalPath: repoInfo["localPath"],
		}
		initConfig.Repos = append(initConfig.Repos, repoConfig)
	}

	// 保存到文件
	data, err := json.MarshalIndent(initConfig, "", "  ")
	if err != nil {
		writeLog(fmt.Sprintf("ERROR: Failed to marshal JSON: %v", err))
		os.Exit(1)
	}

	if err := ioutil.WriteFile(jsonPath, data, 0644); err != nil {
		writeLog(fmt.Sprintf("ERROR: Failed to save JSON file: %v", err))
		os.Exit(1)
	}

	writeLog(fmt.Sprintf("Init config saved to: %s", jsonPath))
	writeLog("Config content:")
	fmt.Println(string(data))
}

// 处理sync模式
func processSyncMode(csvData []map[string]string, jsonPath string, targetPath string) {
	// 读取现有JSON文件
	syncConfig := SyncConfig{}
	syncConfig.TargetPath = targetPath
	syncConfig.Repos = []SyncRepoConfig{}

	if _, err := os.Stat(jsonPath); err == nil {
		writeLog(fmt.Sprintf("Reading existing JSON file: %s", jsonPath))
		data, err := ioutil.ReadFile(jsonPath)
		if err != nil {
			writeLog(fmt.Sprintf("WARNING: Failed to read existing JSON file: %v", err))
		} else {
			if err := json.Unmarshal(data, &syncConfig); err != nil {
				writeLog("WARNING: Invalid JSON file, creating new configuration")
				syncConfig = SyncConfig{}
				syncConfig.TargetPath = targetPath
				syncConfig.Repos = []SyncRepoConfig{}
			}
		}
	} else {
		writeLog("JSON file does not exist, creating new configuration")
	}

	// 创建repo map
	repoMap := make(map[string][]string)
	repoNameMap := make(map[string]string)
	for _, repo := range syncConfig.Repos {
		repoMap[repo.RepoPath] = repo.SkillNames
		repoNameMap[repo.RepoPath] = repo.RepoName
	}

	modified := false

	// 处理CSV数据
	for _, row := range csvData {
		repoPath := strings.TrimSpace(row["repoPath"])
		skillName := strings.TrimSpace(row["skillName"])
		repoName := strings.TrimSpace(row["repoName"])

		if repoPath == "" || skillName == "" {
			writeLog(fmt.Sprintf("Skipping empty row: repoPath=%s, skillName=%s", repoPath, skillName))
			continue
		}

		if _, ok := repoMap[repoPath]; ok {
			skillNames := repoMap[repoPath]
			found := false
			for _, s := range skillNames {
				if s == skillName {
					found = true
					break
				}
			}
			if !found {
				writeLog(fmt.Sprintf("Adding skill '%s' to existing repo: %s", skillName, repoPath))
				skillNames = append(skillNames, skillName)
				repoMap[repoPath] = skillNames
				modified = true
			} else {
				writeLog(fmt.Sprintf("Skill '%s' already exists in repo: %s", skillName, repoPath))
			}
		} else {
			writeLog(fmt.Sprintf("Creating new repo config: %s, adding skill: %s", repoPath, skillName))
			skillNames := []string{skillName}
			repoMap[repoPath] = skillNames
			if repoName != "" {
				repoNameMap[repoPath] = repoName
			}
			modified = true
		}
	}

	// 更新repos
	newRepos := []SyncRepoConfig{}
	for repoPath, skillNames := range repoMap {
		repo := SyncRepoConfig{
			RepoPath:   repoPath,
			RepoName:   repoNameMap[repoPath],
			SkillNames: skillNames,
		}
		newRepos = append(newRepos, repo)
	}
	syncConfig.Repos = newRepos

	// 保存JSON
	if modified {
		data, err := json.MarshalIndent(syncConfig, "", "  ")
		if err != nil {
			writeLog(fmt.Sprintf("ERROR: Failed to marshal JSON: %v", err))
			os.Exit(1)
		}

		if err := ioutil.WriteFile(jsonPath, data, 0644); err != nil {
			writeLog(fmt.Sprintf("ERROR: Failed to save JSON file: %v", err))
			os.Exit(1)
		}

		writeLog(fmt.Sprintf("JSON config saved to: %s", jsonPath))
		writeLog("Config content:")
		fmt.Println(string(data))
	} else {
		writeLog("No updates needed")
	}
}

// 写入日志
func writeLog(message string) {
	timestamp := time.Now().Format(dateFormat)
	fmt.Printf("[%s] %s\n", timestamp, message)
}
